File: implementations/javascript/mergesort.js
function merge(left, right, arr) {
    var a = 0;

    while (left.length && right.length) {
        arr[a++] = (right[0] < left[0]) ? right.shift() : left.shift();
    }
    while (left.length) {
        arr[a++] = left.shift();
    }
    while (right.length) {
        arr[a++] = right.shift();
    }
}

function mergeSort(arr) {
    var len = arr.length;

    if (len === 1) { return; }

    var mid = Math.floor(len / 2),
        left = arr.slice(0, mid),
        right = arr.slice(mid);

    mergeSort(left);
    mergeSort(right);
    merge(left, right, arr);
}

function customRandom(x)
{
    var maxi=1000000; 
    return Math.floor(x*x*x*Math.cos(x)*Math.cos(x)) % maxi ; 
}


/***************************getting arguments ***************************/
var n = (process.argv.length > 2 ) ? parseInt(process.argv[2]):100; 
var niters = (process.argv.length > 3 ) ? parseInt(process.argv[3]):1; 
/***************************main loop ***************************/
for (var iter = 0; iter < niters; iter++) {
    var arr =[];
    for (var i = 0; i < n; i++) {
        arr[i] = customRandom(i); 
    }
    
    mergeSort(arr); 
      
}

File: implementations/ruby/mergesort.rb
def merge_sort(m)
    return m if m.length <= 1
    middle = m.length / 2
    left = merge_sort(m[0...middle])
    right = merge_sort(m[middle..-1])
    merge(left, right)
end
   
def merge(left, right)
	result = []
	until left.empty? || right.empty?
		result << (left.first<=right.first ? left.shift : right.shift)
	end
	result + left + right
end
   
def customRandom(x)
	maxi = 1000000
	(x*x*x*Math.cos(x)*Math.cos(x)).floor % maxi	
end

n = ARGV.length >0 ? ARGV[0].to_i : 20 
niters = ARGV.length >1 ? ARGV[1].to_i : 1 
  
for iter in 1..niters 
	arr = []
	for i in 1..n 
		arr << customRandom(i)
	end 
	p arr 
	p merge_sort(arr)                  #

end 
